k2/csrc/thread_pool.cu

```diff
@@ -0,0 +1,94 @@
+/**
+ * @brief Thread pool.
+ *
+ *
+ * @copyright
+ * Copyright (c)  2020  Xiaomi Corporation (authors: Fangjun Kuang)
+ *
+ * @copyright
+ * See LICENSE for clarification regarding multiple authors
+ */
+
+#include <utility>
+
+#include "k2/csrc/thread_pool.h"
+
+namespace k2 {
+
+static int32_t GetDefaultNumThreads() {
+  int num_threads = std::thread::hardware_concurrency();
+  if (num_threads == 0) num_threads = 1;
+  return num_threads;
+}
+
+ThreadPool::ThreadPool(int32_t num_threads)
+    : threads_(num_threads > 0 ? num_threads : GetDefaultNumThreads()) {
+  for (auto &thread : threads_)
+    thread = std::thread([this]() { this->ProcessTasks(); });
+}
+
+ThreadPool::~ThreadPool() {
+  {
+    std::lock_guard<std::mutex> lock(mutex_);
+    keep_running_ = false;
+    not_empty_cond_.notify_all();  // wake up all threads in the pool
+  }
+  for (auto &thread : threads_) thread.join();
+}
+
+void ThreadPool::WaitAllTasksFinished() {
+  std::unique_lock<std::mutex> lock(mutex_);
+  while (!finished_) {
+    // wait for the `empty_cond_` condition.
+    empty_cond_.wait(lock);
+  }
+}
+
+void ThreadPool::ProcessTasks() {
+  std::unique_lock<std::mutex> lock(mutex_);
+  while (keep_running_) {
+    while (tasks_.empty() && keep_running_) {
+      // Wait for the `not_empty_` condition.
+      // RunTask() or the destructor will signal it.
+      not_empty_cond_.wait(lock);
+    }
+
+    if (!keep_running_) break;
```

Comment:
I think it's not right here as we may leave this task un-executed, but we have wait for it with above mutex, according to the documentation, we should return after execute it.

Reply:
The behavior is expected.
https://github.com/k2-fsa/k2/blob/6df0cd0a53ffc892cf04cd55e4f6be8ace99ff0d/k2/csrc/thread_pool.h#L37-L41